### PR TITLE
[MAINT] using tox to test in different environments

### LINKED ISSUE #3718: cancel concurrent CI workflows ?
Some CI workflows (testing, doc build) take a long time to run.

Those are set to keep running even if new commits are pushed on a branch / PR that can render the end result of that workflow obsolete.

This also means that compulsive "git pusher" (I am one of them 🙈) tend to end up with many CI workflows queued for a given PR, which hogs resources and also potentially delays the start of workflows from other PRs / devs.

I'd be tempted to cancel concurrent previous workflows for testing and doc, the way it is done with the black / flake8 / isort workflows: https://github.com/nilearn/nilearn/blob/d50ee5201c1ad7f5d583c5c4d0ec0ea6d8ef6a5d/.github/workflows/isort.yml#L12

Wondering how others feel about this.

### LINKED ISSUE #4019: Try using Tox to simplify testing with different environments
- could simplify the pyproject.toml 
- could simplify CI workflows by just calling the right tox recipe rather than setting environment variables and calling some bash script
- could move some of the recipes from the makefile to the tox.ini

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #4019
- Closes #3718
- Relates to #3812 

Changes proposed in this pull request:

- define all tests environments in tox.ini and not pyproject.toml
- define linting environment in tox.ini
- channel linting via pre-commit
- move test, test-doc and lint "recipes" from makefile and bash scripts to tox.ini
- call tox in CI to run tests and flake8
- drop python 3.7 in CI
- do not rely on conda during the build of the doc

### TODO

- [x] run doc tests (all OS and all python versions)
- [x] take care of coverage
- [x] only run mininum tests on ubuntu
- [x] update CI README
- [x] update documentation

### CODE REVIEW

**Comment by Remi-Gau** [comment]
Wonder if some minimal documentation should be added in contributing or maintenance regarding the use of tox? @ymzayek @bthirion

File: .github/workflows/build-docs.yml
Comment (Remi-Gau) [inline .github/workflows/build-docs.yml:107]: @ymzayek this is after this step you mention we can skip the cache restore if we are on a partial build?

File: tox.ini
Comment (ymzayek) [inline tox.ini:55]: should plotly contain mpl dependency? if so for the way you've set it up maybe change the name of this section to plotting

**Comment by ymzayek** [comment]
> Wonder if some minimal documentation should be added in contributing or maintenance regarding the use of tox?

I would start by adding something in the maintenance docs and we can see if something would fit for the contributing or at least reference the new section for people who might be interested. Also document in the workflow readme

File: tox.ini
Comment (Remi-Gau) [reply tox.ini:55]: makes sense

File: .github/workflows/build-docs.yml
Comment (ymzayek) [reply .github/workflows/build-docs.yml:107]: Yes so we can skip the caching steps if `BUILD` is set to `html-noplot`

File: .github/workflows/build-docs.yml
Comment (ymzayek) [inline .github/workflows/build-docs.yml:97]: Move this and the Build docs step commands below to tox as well?

**Comment by Remi-Gau** [comment]
Interesting. We are getting 3 tests to fail only on windows that never showed up before:

https://github.com/nilearn/nilearn/actions/runs/6410072517/job/17402608737#step:5:15020

```
FAILED nilearn/surface/tests/test_surface.py::test_load_surf_mesh_file_freesurfer
FAILED nilearn/surface/tests/test_surface.py::test_load_surf_mesh_file_error
FAILED nilearn/surface/tests/test_surface.py::test_load_surf_mesh_file_glob
```

File: .github/workflows/build-docs.yml
Comment (Remi-Gau) [reply .github/workflows/build-docs.yml:97]: Not sure I want to move more of the doc build under tox.

Reading through the FAQ I can imagine some potential path headaches in CI.

https://tox.wiki/en/4.11.3/faq.html#building-documentation-with-sphinx

File: .github/workflows/README.md
Comment (ymzayek) [inline .github/workflows/README.md:19]: why change the paths? it takes nilearn main as root

**Comment by Remi-Gau** [comment]
> Interesting. We are getting 3 tests to fail only on windows that never showed up before:
> 
> https://github.com/nilearn/nilearn/actions/runs/6410072517/job/17402608737#step:5:15020
> 
> ```
> FAILED nilearn/surface/tests/test_surface.py::test_load_surf_mesh_file_freesurfer
> FAILED nilearn/surface/tests/test_surface.py::test_load_surf_mesh_file_error
> FAILED nilearn/surface/tests/test_surface.py::test_load_surf_mesh_file_glob
> ```
> 
> Error

some errors come from the username not being set. this is kind of strange.

```
    def getuser():
        """Get the username from the environment or password database.
    
        First try various environment variables, then the password
        database.  This works on Windows as long as USERNAME is set.
    
        """
    
        for name in ('LOGNAME', 'USER', 'LNAME', 'USERNAME'):
            user = os.environ.get(name)
            if user:
                return user
    
        # If this fails, the exception will "explain" why
```

**Comment by Remi-Gau** [comment]
> Currently leads to duplication of where the dependencies for style checking (black, flake...) are declared (in tox.ini and pyproject.toml).

possible work around for this may be to run `pre-commit run -a` via tox

File: .github/workflows/build-docs.yml
Comment (Remi-Gau) [inline .github/workflows/build-docs.yml:126]: @ymzayek this is to skip data caching: makes sense to you?

Comment (ymzayek) [reply .github/workflows/build-docs.yml:126]: Yes perfect!

File: doc/maintenance.rst
Comment (Remi-Gau) [inline doc/maintenance.rst:133]: @ymzayek 

how does this look in terms of doc?

File: nilearn/conftest.py
Comment (ymzayek) [inline nilearn/conftest.py:336]: Just to be clear, is the error that warrants this related to this PR? If so, can you explain?
Haven't seen it on main or in other PRs. More just for my understanding...

File: doc/maintenance.rst
Comment (ymzayek) [inline doc/maintenance.rst:146]:
```suggestion
You can set up certain environment or run certain command by calling ``tox``.
```

**Review by ymzayek — COMMENTED** [review]
Looking good!

File: nilearn/conftest.py
Comment (Remi-Gau) [reply nilearn/conftest.py:336]: not seen in any other PR either

so far I was trying to solve this to at least make sure that my assumptions about what the problem is

Comment (Remi-Gau) [reply nilearn/conftest.py:336]: Seems related to this: https://github.com/tox-dev/tox/issues/1455

May be easier to solve in the tox env definition: let me check.

Comment (Remi-Gau) [reply nilearn/conftest.py:336]: ok so you can pass env variables in the tox config and this solves the problem: good thing you pushed me to look further

File: tox.ini
Comment (ymzayek) [inline tox.ini:1]: very minor thing: irregular indentations under `commands` paramter

File: .github/workflows/test_with_tox.yml
Comment (ymzayek) [inline .github/workflows/test_with_tox.yml:1]: shouldn't we run coverage and upload to Codecov for prerelease and min_reqs as well? I'm not sure if it's the case but I'm thinking if there are some lines that run only in those particular environments, they will not be in the Codecov report

File: .github/workflows/test_with_tox.yml
Comment (ymzayek) [inline .github/workflows/test_with_tox.yml:20]:
```suggestion
        name: Test with ${{ matrix.py }} on ${{ matrix.os }}
```

File: .github/workflows/test_with_tox.yml
Comment (Remi-Gau) [reply .github/workflows/test_with_tox.yml:1]: Yeah I can set that up. At least we will know if we get different coverage.

Also if we do that it probably means we can refactor the workflow to have a single job with a matrix with python version, os, environment, env variable.

Comment (ymzayek) [reply .github/workflows/test_with_tox.yml:1]: Yea at least that's my understanding of why the old set up did this. I guess it's easiest just to add it as a last step to each workflow and have codecov take care of merging the reports. But if you feel like refactoring it go for it.

**Comment by bthirion** [comment]
> Wonder if some minimal documentation should be added in contributing or maintenance regarding the use of tox? @ymzayek @bthirion

I thin so. It's actually unclear to me what the impact for contributors is.

File: doc/maintenance.rst
Comment (bthirion) [inline doc/maintenance.rst:176]: The explanation is pretty clear, thx.
I wonder whether this covers all the use-cases a maintainer/developer may encounter.

**Review by bthirion — COMMENTED** [review]
LGTM otherwise.

**Review by ymzayek — APPROVED** [review]
LGTM thx @Remi-Gau !

**Comment by ymzayek** [comment]
I guess we should hold off merging anyways until we discuss dropping 3.7 support

**Comment by Remi-Gau** [comment]
> I guess we should hold off merging anyways until we discuss dropping 3.7 support

switching to draft to avoid too early a merge then

File: .github/workflows/build-docs.yml
Comment (ymzayek) [inline .github/workflows/build-docs.yml:123]:
```suggestion
                if [[ "$(cat build.txt)" == "ci-html-noplot" ]]; then
```

**Comment by Remi-Gau** [comment]
# TODO

make sure coverage is uploaded

current status:

```
     _____          _
    / ____|        | |
   | |     ___   __| | ___  ___ _____   __
   | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
   | |___| (_) | (_| |  __/ (_| (_) \ V /
    \_____\___/ \__,_|\___|\___\___/ \_/

  Codecov report uploader 0.6.3
[2023-10-12T06:52:18.089Z] ['info'] => Project root located at: /home/runner/work/nilearn/nilearn
[2023-10-12T06:52:18.094Z] ['info'] -> No token specified or token is empty
[2023-10-12T06:52:18.102Z] ['info'] Searching for coverage files...
[2023-10-12T06:52:18.236Z] ['info'] Warning: Some files located via search were excluded from upload.
[2023-10-12T06:52:18.236Z] ['info'] If Codecov did not locate your files, please review https://docs.codecov.com/docs/supported-report-formats
[2023-10-12T06:52:18.236Z] ['error'] There was an error running the uploader: No coverage files located, please try use `-f`, or change the project root with `-R`
```

**Comment by Remi-Gau** [comment]
Progress but not success

```
_____          _
    / ____|        | |
   | |     ___   __| | ___  ___ _____   __
   | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
   | |___| (_) | (_| |  __/ (_| (_) \ V /
    \_____\___/ \__,_|\___|\___\___/ \_/

  Codecov report uploader 0.7.0
[2023-10-19T09:37:19.900Z] ['info'] => Project root located at: D:/a/nilearn/nilearn
[2023-10-19T09:37:19.904Z] ['info'] -> No token specified or token is empty
[2023-10-19T09:37:19.950Z] ['info'] Searching for coverage files...
[2023-10-19T09:37:20.056Z] ['info'] Warning: Some files located via search were excluded from upload.
[2023-10-19T09:37:20.056Z] ['info'] If Codecov did not locate your files, please review https://docs.codecov.com/docs/supported-report-formats
[2023-10-19T09:37:20.057Z] ['info'] => Found 1 possible coverage files:
  coverage.xml
[2023-10-19T09:37:20.057Z] ['info'] Processing D:/a/nilearn/nilearn/coverage.xml...
[2023-10-19T09:37:20.072Z] ['info'] Detected GitHub Actions as the CI provider.
[2023-10-19T09:37:20.935Z] ['info'] Pinging Codecov: https://codecov.io/upload/v4?package=github-action-3.1.4-uploader-0.7.0&token=*******&branch=tox&build=6572360256&build_url=https%3A%2F%2Fgithub.com%2Fnilearn%2Fnilearn%2Factions%2Fruns%2F6572360256&commit=497ac61046538f6f0e98cd7e1ef4be495ef30e6f&job=test&pr=4029&service=github-actions&slug=nilearn%2Fnilearn&name=&tag=&flags=windows-latest_3.12_test_plotting&parent=
[2023-10-19T09:37:21.195Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
```

**Comment by ymzayek** [comment]
Now there are 2 checks results for codecov/patch and 2 for codecov/project. I see that the Details link for each respectively bring you either to github checks or directly to codecov but why not just link to one? I personally prefer the github checks for the quick summary and then from there you can easily open up codecov.

**Comment by Remi-Gau** [comment]
> Now there are 2 checks results for codecov/patch and 2 for codecov/project. I see that the Details link for each respectively bring you either to github checks or directly to codecov but why not just link to one? I personally prefer the github checks for the quick summary and then from there you can easily open up codecov.

this is new

I have never seen this before

File: tox.ini
Comment (bthirion) [inline tox.ini:123]: Not sure I understood. What is meant here is that
`test_plot_min == test_min without plotly`  
?

**Review by bthirion — APPROVED** [review]
LGTM, thx !

File: tox.ini
Comment (ymzayek) [reply tox.ini:123]: Plotly is additional for supporting interactive plots but is not actually needed to use nilearn's plotting functionality. So to me it makes sense not to be included in the minimum versions for plotting. This could be made more clear in the description perhaps.

**Review by ymzayek — APPROVED** [review]
Other than the codecov comment this PR looks good to me!

**Comment by ymzayek** [comment]
Perhaps we should merge this and check this https://github.com/nilearn/nilearn/pull/4029#issuecomment-1779132041 separately?

**Comment by Remi-Gau** [comment]
> Perhaps we should merge this and check this [#4029 (comment)](https://github.com/nilearn/nilearn/pull/4029#issuecomment-1779132041) separately?

Will fix the merge conflict.

I have not had the time to figure out where this double coverage report is coming from. But definitely OK delegating that to a follow up PR.

File: pyproject.toml
Comment (Remi-Gau) [inline pyproject.toml:71]: Was just thinking that changing the name of that list of optional dependencies may "break" some users workflow.

Maybe better to have an "alias" like:

```
plotly = ["nilearn[plotting]"]
```

Though not technically API, not sure if we should not throw some sort of deprecation warning

Comment (Remi-Gau) [reply pyproject.toml:71]: In any case this sounds like it is worth mentioning in the changelog

Comment (Remi-Gau) [reply pyproject.toml:71]: or maybe just keep the original "plotting" and "plotly" original names and content...

Comment (Remi-Gau) [reply pyproject.toml:71]: OK: sounds reasonable

I will still add the option: `plotly = ["nilearn[plotting]"]` because this won't lead to any breakage.

The 'only' behavioral change this will do is that the `plotting` option will now also install plotly and kaleido.

File: tox.ini
Comment (ymzayek) [inline tox.ini:49]: Hm if this doesn't work try `kaleido==0.1.0.post1 ; sys_platform == 'windows'`

the example they provide is `pywin32 >=1.0 ; sys_platform == 'win32'` so not exactly the same but maybe it understands both idk

Comment (Remi-Gau) [reply tox.ini:49]: I thought the syntax would be the same as pyproject.toml but silly me to expect uniformity in the python packaging world...

Comment (ymzayek) [reply tox.ini:49]: well looks like it worked actuallt!

**Comment by Remi-Gau** [comment]
Ok will update the changelog and I think this should rlbe ready to go in.

File: doc/changes/latest.rst
Comment (Remi-Gau) [inline doc/changes/latest.rst:33]: filing this under API... but not sure this is the best

Comment (ymzayek) [reply doc/changes/latest.rst:33]: why not :bdg-secondary:\`Maint\` ?

Comment (Remi-Gau) [reply doc/changes/latest.rst:33]: oh dear...
because I had forgotten we had it: fixed now

Comment (ymzayek) [reply doc/changes/latest.rst:33]: Did you push the fix?

**Comment by Remi-Gau** [comment]
ok merging then

File: doc/changes/latest.rst
Comment (Remi-Gau) [reply doc/changes/latest.rst:33]: I thought I had...

Comment (Remi-Gau) [reply doc/changes/latest.rst:33]: I guess this way we can test the whole tox thing on a PR...
https://github.com/nilearn/nilearn/pull/4114

**Pull request merged**
